dialects/direct/passes/ds2cps.cpp

```diff
@@ -0,0 +1,266 @@
+            }
+
+            // continuation of call site to receive result
+            auto fun_cont = world.nom_lam(world.cn(ret_ty), world.dbg(currentLambda->name() + "_cont"));
+            fun_cont->set_filter(currentLambda->filter());
+
+            // f a -> f_cps(a,cont)
+            DefArray ext_args(args.size() + 1, [&](size_t i) {
+                if (i == args.size()) {
+                    return (const Def*)fun_cont;
+                } else {
+                    return args[i];
+                }
+            });
+#ifdef verbose_rewrite
+            std::cout << "  cps call " << lam_cps;
+            std::cout << " : " << lam_cps->type() << " with ";
+            std::cout << world.tuple(ext_args) << " : " << world.tuple(ext_args)->type() << std::endl;
+#endif
+            auto cps_call = world.app(lam_cps, ext_args, world.dbg("cps_call"));
+            currentLambda->set_body(cps_call);
+            currentLambda->set_filter(true);
+
+#ifdef verbose_rewrite
+            std::cout << "  overwrote body of " << currentLambda << " : " << currentLambda->type() << " with "
+                      << cps_call << " : " << cps_call->type() << std::endl;
+#endif
+
+            // write the body context in the newly created continuation
+            // that has access to the result (as its argument)
+            currentLambda = fun_cont;
+            // result of ds function
+            auto [res] = fun_cont->vars<1>();
+            // fun_cont->set_body(world.bot());
+
+#ifdef verbose_rewrite
+            std::cout << "  result " << res << " : " << res->type() << " instead of " << def << " : " << def->type()
+                      << std::endl;
+#endif
+            // replace call with the result in the context that will be placed in the continuation
+            return res;
+        }
+
+        // TODO:
+        // are ops rewrites
+        // + app calle/arg rewrites
+        // all possible combinations?
+
+        // non lam call (handled above)
+        auto arg            = app->arg();
+        auto args_rewritten = rewrite_(arg);
+        auto res            = world.app(rewrite_(callee), args_rewritten);
+        return res;
+    }
+
+    // TODO: check if lam is necessary or var is enough
+    if (auto var = def->isa<Var>()) {
+        // do not descend into infinite chain through function
+        return def;
+    }
+
+    for (size_t i = 0, e = def->num_ops(); i != e; ++i) {
+        auto op  = def->op(i);
+        auto nop = rewrite_(op);
+#ifdef verbose_rewrite
+        std::cout << "  def " << def << " : " << def->type() << " node " << def->node_name() << std::endl;
+        std::cout << "  refine " << op << " : " << op->type() << " to (" << nop << ") : " << nop->type() << " ["
+                  << nop->node_name() << "]" << std::endl;
+#endif
+        auto ndef = def->refine(i, nop);
+        def       = ndef;
+    }
+
+    return def;
+}
+
+PassTag* DS2CPS::ID() {
+    static PassTag Key;
+    return &Key;
+}
+
```

Comment:
can you also format with `clang-format`? There are some odd formattings here and there.

Reply:
I forgot. 
:+1: 
I should probably enable format on save.